src/celeritas/optical/interactor/WavelengthShiftInteractor.hh

```diff
@@ -0,0 +1,164 @@
+
+    // Incident photon energy
+    Energy const inc_energy_;
+    // Sample distributions
+    PoissonDistribution<real_type> sample_num_photons_;
+    ExponentialDistribution<real_type> sample_time_;
+    // Grid calculators
+    GenericCalculator calc_energy_;
+    GenericCalculator calc_cdf_;
+    // Allocate space for secondary particles
+    SecondaryAllocator& allocate_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION
+WavelengthShiftInteractor::WavelengthShiftInteractor(
+    WLSDataCRef const& shared,
+    ParticleTrackView const& particle,
+    OpticalMaterialId const& mat_id,
+    SecondaryAllocator& allocate)
+    : inc_energy_(particle.energy())
+    , sample_num_photons_(shared.wls_scalars[mat_id].mean_num_photons)
+    , sample_time_(real_type{1} / shared.wls_scalars[mat_id].time_constant)
+    , calc_energy_(shared.energy_cdf[mat_id], shared.reals)
+    , calc_cdf_(GenericCalculator::from_inverse(shared.energy_cdf[mat_id],
+                                                shared.reals))
+    , allocate_(allocate)
+{
+    CELER_EXPECT(inc_energy_.value() > calc_energy_(0));
+}
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the wavelength shift (WLS) photons as in G4OpWLS of the Geant4
+ * release 11.2.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction WavelengthShiftInteractor::operator()(Engine& rng)
+{
+    // Sample the number of photons generated from WLS
+    unsigned int num_photons = sample_num_photons_(rng);
+
+    if (num_photons == 0)
+    {
+        // Return unchanged photon without reemitted optical photons
+        return Interaction::from_unchanged();
+    }
+
+    // Allocate space for reemitted optical photons
+    TrackInitializer* secondaries = allocate_(num_photons);
+
+    if (secondaries == nullptr)
+    {
+        // Failed to allocate space
+        return Interaction::from_failure();
+    }
+
+    // Sample wavelength shifted optical photons
+    Interaction result = Interaction::from_absorption();
+    result.secondaries = {secondaries, num_photons};
+
+    IsotropicDistribution sample_direction{};
+    for (size_type i : range(num_photons))
+    {
+        // Sample the emitted energy from the inverse cumulative distribution
+        real_type energy = calc_energy_(generate_canonical(rng));
+        if (CELER_UNLIKELY(energy > inc_energy_.value()))
+        {
+            // Sample a restricted energy below the incident photon energy
+            real_type cdf_max = calc_cdf_(inc_energy_.value());
+            UniformRealDistribution<real_type> sample_cdf(0, cdf_max);
+            energy = calc_energy_(sample_cdf(rng));
```

Comment:
I think we need a CDF sampler data type + class, rather than constructing an inverted xs/cdf data, then creating an xs calculator, and having these two helper clasess.

Reply:
Thanks, it would be great if you can help to write a CDF sampler class.

Reply:
For muon pair production we also need to sample from a CDF, but on a 2D grid like we do in the [cascade collider](https://github.com/celeritas-project/celeritas/blob/e390d55956c3f59124f97bf7cddb8d1d259ecd93/src/celeritas/neutron/interactor/detail/CascadeCollider.hh#L146-L161): I was also thinking it would be nice to have a helper class for that.

Reply:
I added a to-do here; we don't *have* to have this as part of the PR.